File: algorithms/easy/remove-duplicates-from-sorted-list.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x, nt=None):
        self.val = x
        self.next = nt


class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        pre = ListNode(None)
        pre.next = head
        cur = head
        head = pre
        dic = {}
        while pre and pre.next:
            if cur and dic.get(cur.val):
                pre.next = cur.next
            else:
                pre = pre.next
                dic[cur.val] = 1
            cur = cur.next
        return head.next


def test():
    head = ListNode(1, ListNode(2, ListNode(1, ListNode(2))))
    Solution().deleteDuplicates(head)
    l = []
    while head:
        l.append(head.val)
        head = head.next
    assert l == [1, 2]

if __name__ == '__main__':
    test()

File: algorithms/easy/remove-duplicates-from-sorted-list.c
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* newNode(int val, struct ListNode *next) {
    struct ListNode *node = malloc(sizeof(struct ListNode));
    node->val = val;
    node->next = next;
    return node;
}

struct ListNode* deleteDuplicates(struct ListNode* head) {
    if (head == NULL)
        return NULL;
    struct ListNode *cur = head;
    while (cur != NULL) {
        if (cur->next != NULL && cur->val == cur->next->val) {
            cur->next = cur->next->next;
        } else {
            cur = cur->next;
        }
    }
    return head;
}

int main() {
    struct ListNode *l1 = newNode(2, newNode(2, newNode(2, newNode(2, newNode(2, NULL)))));
    struct ListNode *l2 = deleteDuplicates(l1);
    while(l2 != NULL) {
        printf("%d ", l2->val);
        l2 = l2->next;
    }
    printf("\n");
}
